### PR TITLE
feat(server): add DELETE API for Carapace certificates

### LINKED ISSUE #454: Implement API Create/Delete and edit certificate
API for create/deleted and edit certificate

### PR DESCRIPTION
**Pull request by NiccoMlt**
Closes #454

### CODE REVIEW

**Review by hamadodene — APPROVED** [review]
LGTM...
@NiccoMlt  Can you check the test that fails? It could be one of the flacky tests. If you can fix it it wouldn't be bad.

**Comment by hamadodene** [comment]
Are you planning to integrate it on the admin side too?
Otherwise, can you make another follow up issue to integrate it if it isn't already there?

**Comment by NiccoMlt** [comment]
> Are you planning to integrate it on the admin side too? Otherwise, can you make another follow up issue to integrate it if it isn't already there?

The issue already exists, the UI support was not planned for this issue: https://github.com/diennea/carapaceproxy/issues/357

> LGTM... @NiccoMlt Can you check the test that fails? It could be one of the flacky tests. If you can fix it it wouldn't be bad.

Ok, I'll give it a look

File: carapace-server/src/main/java/org/carapaceproxy/api/CertificatesResource.java [hunk 1]
```diff
@@ -261,25 +269,40 @@ public Response createCertificate(CertificateForm form) {
         return FormValidationResponse.created();
     }
 
+    @DELETE
+    @Path("{certId}")
+    public Response createCertificate(@PathParam("certId") final String certId) {
```
Comment (dmercuriali) [inline carapace-server/src/main/java/org/carapaceproxy/api/CertificatesResource.java:274]: should be `deleteCertificate`

File: carapace-server/src/main/java/org/carapaceproxy/api/CertificatesResource.java [hunk 2]
```diff
@@ -261,25 +269,40 @@ public Response createCertificate(CertificateForm form) {
...
     @Path("{certId}/download")
-    public Response downloadCertificateById(@PathParam("certId") String certId) throws GeneralSecurityException {
+    public Response downloadCertificateById(@PathParam("certId") final String certId) {
         CertificateBean cert = findCertificateById(certId);
-        byte[] data = new byte[0];
-        if (cert != null && cert.isDynamic()) {
-            HttpProxyServer server = (HttpProxyServer) context.getAttribute("server");
-            DynamicCertificatesManager dynamicCertificateManager = server.getDynamicCertificatesManager();
-            data = dynamicCertificateManager.getCertificateForDomain(cert.getId());
+        if (cert == null || !cert.isDynamic()) {
```
Comment (dmercuriali) [inline carapace-server/src/main/java/org/carapaceproxy/api/CertificatesResource.java:293]: This is a behaviour change. I'm ok with this but next time It would be best to create another PR.

We must remember to write this change in the release notes

File: carapace-server/src/main/java/org/carapaceproxy/configstore/HerdDBConfigurationStore.java [hunk 3]
```diff
@@ -475,6 +494,21 @@ public void saveCertificate(CertificateData cert) {
         }
     }
 
+    @Override
+    public void removeCertificate(final String certId) {
+        try {
+            final var connection = datasource.getConnection();
+            final var preparedStatement = connection.prepareStatement(REMOVE_DIGITAL_CERTIFICATES_TABLE);
```
Comment (dmercuriali) [inline carapace-server/src/main/java/org/carapaceproxy/configstore/HerdDBConfigurationStore.java:501]: why not opening both in the twr? if the `prepareStatement` call fails the connection is not closed

**Review by dmercuriali — CHANGES_REQUESTED** [review]

File: carapace-server/src/main/java/org/carapaceproxy/api/CertificatesResource.java (hunk 1 above)
Comment (NiccoMlt) [reply carapace-server/src/main/java/org/carapaceproxy/api/CertificatesResource.java:274]: Argh, you're right!

File: carapace-server/src/main/java/org/carapaceproxy/api/CertificatesResource.java (hunk 2 above)
Comment (NiccoMlt) [reply carapace-server/src/main/java/org/carapaceproxy/api/CertificatesResource.java:293]: Actually, it only changed the error from a non-meaningful 500 originated from a NullPointerException to a more meaningful 404.

But you're right, I should've stated the updated somewhere

File: carapace-server/src/main/java/org/carapaceproxy/configstore/HerdDBConfigurationStore.java (hunk 3 above)
Comment (NiccoMlt) [reply carapace-server/src/main/java/org/carapaceproxy/configstore/HerdDBConfigurationStore.java:501]: The try-with-resources is in the line below: from Java 9 it should be an alternatively valid syntax

Comment (dmercuriali) [reply carapace-server/src/main/java/org/carapaceproxy/configstore/HerdDBConfigurationStore.java:501]: It's valid syntax but I think the compiled code does not have the same meaning. By declaring the resources outside the twr, if the call to `connection#prepareStatement` fails then `connection#close` is not invoked.

PS. actually tested out of curiosity and I can confirm this. With this syntax the 2 resources are declared and then the 2 nested try are created. By declaring both inside the twr you end up with the first declaration and the first try block, which wraps the second declaration and the second try block

Comment (NiccoMlt) [reply carapace-server/src/main/java/org/carapaceproxy/configstore/HerdDBConfigurationStore.java:501]: Cool, didn't know that, I totally missed the point of that addition, thank you!

I also tried it like that:

```java
public class Test {

    public interface Common extends AutoCloseable {
        @Override
        default void close() {
            System.out.println(this.getClass() + " closed correctly");
        }

        default Successful getSuccessful() {
            return new Successful();
        }
        
        default Failure getFailure() {
            return new Failure();
        }
    }

    public static class Successful implements Common {
        public Successful() {
            System.out.println(this.getClass() + " created correctly");
        }
    }

    public static class Failure implements Common {
        public Failure() {
            System.out.println(this.getClass() + " created");
            throw new IllegalStateException("Faiulre");
        }
    }

    public static void classic() {
        try (final var succ = new Successful(); final var fail = succ.getFailure()) {
            System.out.println("Try block");
        }
    }

    public static void java9() {
        final var succ = new Successful();
        try (final var fail = succ.getFailure()) {
            System.out.println("Try block");
        }
    }

    public static void likeIDid() {
        final var succ = new Successful();
        final var fail = succ.getFailure();
        try (succ; fail) {
            System.out.println("Try block");
        }
    }

    public static void main(final String[] args) {
        System.out.println("classic...");
        try {
            classic();
        } catch (Exception ignored) { }
        System.out.println("Java 9...");
        try {
            java9();
        } catch (Exception ignored) { }
        System.out.println("Like I did...");
        try {
            likeIDid();
        } catch (Exception ignored) { }
    }
}
```

which is compiled like that:

```java
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by FernFlower decompiler)
//

public class Test {
    public Test() {
    }

    public static void classic() {
        Successful var0 = new Successful();

        try {
            Failure var1 = var0.getFailure();

            try {
                System.out.println("Try block");
            } catch (Throwable var6) {
                if (var1 != null) {
                    try {
                        var1.close();
                    } catch (Throwable var5) {
                        var6.addSuppressed(var5);
                    }
                }

                throw var6;
            }

            if (var1 != null) {
                var1.close();
            }
        } catch (Throwable var7) {
            try {
                var0.close();
            } catch (Throwable var4) {
                var7.addSuppressed(var4);
            }

            throw var7;
        }

        var0.close();
    }

    public static void java9() {
        Successful var0 = new Successful();
        Failure var1 = var0.getFailure();

        try {
            System.out.println("Try block");
        } catch (Throwable var5) {
            if (var1 != null) {
                try {
                    var1.close();
                } catch (Throwable var4) {
                    var5.addSuppressed(var4);
                }
            }

            throw var5;
        }

        if (var1 != null) {
            var1.close();
        }

    }

    public static void likeIDid() {
        Successful var0 = new Successful();
        Failure var1 = var0.getFailure();
        Successful var2 = var0;

        try {
            Failure var3 = var1;

            try {
                System.out.println("Try block");
            } catch (Throwable var8) {
                if (var1 != null) {
                    try {
                        var3.close();
                    } catch (Throwable var7) {
                        var8.addSuppressed(var7);
                    }
                }

                throw var8;
            }

            if (var1 != null) {
                var1.close();
            }
        } catch (Throwable var9) {
            if (var0 != null) {
                try {
                    var2.close();
                } catch (Throwable var6) {
                    var9.addSuppressed(var6);
                }
            }

            throw var9;
        }

        if (var0 != null) {
            var0.close();
        }

    }

    public static void main(String[] var0) {
        System.out.println("classic...");

        try {
            classic();
        } catch (Exception var4) {
        }

        System.out.println("Java 9...");

        try {
            java9();
        } catch (Exception var3) {
        }

        System.out.println("Like I did...");

        try {
            likeIDid();
        } catch (Exception var2) {
        }

    }

    public static class Successful implements Common {
        public Successful() {
            System.out.println(String.valueOf(this.getClass()) + " created correctly");
        }
    }

    public static class Failure implements Common {
        public Failure() {
            System.out.println(String.valueOf(this.getClass()) + " created");
            throw new IllegalStateException("Faiulre");
        }
    }

    public interface Common extends AutoCloseable {
        default void close() {
            System.out.println(String.valueOf(this.getClass()) + " closed correctly");
        }

        default Successful getSuccessful() {
            return new Successful();
        }

        default Failure getFailure() {
            return new Failure();
        }
    }
}
```

and in fact, if run, gives the behavior that you were hinting:

```
❯ java Test                                                                                                          
classic...
class Test$Successful created correctly
class Test$Failure created
class Test$Successful closed correctly
Java 9...
class Test$Successful created correctly
class Test$Failure created
Like I did...
class Test$Successful created correctly
class Test$Failure created
```

Thank you for pointing it out, I'll fix it

**Comment by NiccoMlt** [comment]
> > LGTM... @NiccoMlt Can you check the test that fails? It could be one of the flacky tests. If you can fix it it wouldn't be bad.
> 
> Ok, I'll give it a look

@hamadodene sorry but I didn't manage to reproduce it in any way, it's easier to break it for "too many files" than to have the same problem happen locally:

image
